### PR TITLE
Support union search space in multivariate `TPESampler` to preserve conditional correlation

### LINKED ISSUE #5299: Improvement to multivariate TPESampler when the union search space is different from the intersection search space
### Motivation

During hyper-parameter optimization I often want to find the best of a few different methods, all of which can have a different set of parameters. To model this with the multivariate TPESampler we're required to use the group decomposition, which samples each group independently of each other. E.g. in a simple objective like the following:
```
def objective(trial):
    x = trial.suggest_categorical("x", ["A", "B"])
    t = trial.suggest_float("t", -2, 2)
    if x == "A":
        y = trial.suggest_float("y", 1, 2)
        return (y - t)**2
    else:
        z = trial.suggest_float("z", -2, 1)
        return (z - t)**2
```
the model will struggle to converge as t is heavily correlated with y and z, but they are modeled independently of each other. This is a very simple example, but highlights some issues with correlation between groups.

### Description

The cleanest way I could think off to avoid the group decomposition is to replace the intersection search space used in the multivariate TPESampler with a union search space. The internal representation of 'missing' parameters can be set to e.g. 'nan' which is then handled in the ParzenEstimator and _MixtureOfProductDistribution. 
During sampling, we can just set a parameter in the sample to 'nan' if the active index had 'nan' values as well. In the likelihood estimate for categorical parameters, we can just introduce 'nan' as an additional category. For integers and floats, we can multiply the univariate distribution with a catagorical one, estimating the likelihood of the parameter being sampled or not.
Of course due to priors, this can result in some required parameters not being sampled, but these can be sampled with an independent tpe sampler.

This doesn't require huge changes in the code, and shouldn't change anything on the user end. If it performs well across tests, it should be able to replace the default {multivariate=True, group=False} behavior, as the behavior is only different from the current one when the intersection search space is different from the union search space.

### Alternatives (optional)

Alternatively we can continue using the group decomposition, but instead determine the parameter hierarchy in the objective. I.e. in the above example, (x, t) is always sampled, so we can draw n_ei_candidates samples for them first using all trials. Then based on their values, we can either sample y or z from the subset of trials that contain these values. This requires some classifier which can either be given as user input, or be determined automatically by a simple decision tree. Then once all the required parameters have been sampled, we can choose the best sample from their collective likelihoods.
This only adds in correlation between the parameters that are part of the objective decision tree, but does help the sampler avoid getting stuck in local minima. However, the changes doesn't mesh well with the TPE sampler code, and would likely have to be implemented as a separate sampler.

### Additional context (optional)

I made implementations of both approaches here: https://github.com/larsbratholm/optuna_tpe_suggestion/tree/main
as well as a benchmark on a slightly more complex objective shown here: https://github.com/larsbratholm/optuna_tpe_suggestion/blob/main/run_benchmark.py#L19

The geometric mean loss over 128 runs for RandomSampler, TPE, multivariate TPE, the non-group multivariate approach suggested, as well as the alternative hierarchical tpe sampler are shown below
benchmark

Both methods perform well on the objective, but the objective features high correlation and a local minima that is easy to get stuck in, so might not be representative overall?

Are either of these approaches something that would be welcome as a contribution to Optuna? And are there other objectives that have been used to benchmark the group decomposed TPESampler that would be helpful to test on?

### PR DESCRIPTION
**Pull request by Rishabh-git10**
## Motivation
Fixes #5299. 
Configuring `TPESampler(multivariate=True, group=False)` on conditional search spaces currently triggers an independent sampling fallback or drops back to intersection group decomposition. This behavior discards parameter observations across different conditional branches, breaking the capability to model correlations between top-level routing choices and nested hyperparameters.

## Description of the changes
This PR implements a union search space approach for multivariate TPE, padding unobserved parameters with `np.nan` so that a single joint distribution can be estimated without losing cross-branch correlation structure.

* **`optuna/search_space/union.py`**: Added `UnionSearchSpace` to extract the full union of parameter distributions suggested across dynamic or conditional trials.
* **`optuna/samplers/_tpe/sampler.py`**: Updated `_get_internal_repr` to pad missing conditional parameters with `np.nan` instead of dropping the trial. Initialized an internal independent fallback sampler within `__init__` to safely resolve missing selection paths without entering infinite recursion cycles.
* **`optuna/samplers/_tpe/parzen_estimator.py`**: Modified `_calculate_categorical_distributions` to dynamically detect `NaN` entries and append an explicit category column representing unobserved parameter states.
* **`optuna/samplers/_tpe/probability_distributions.py`**: Upgraded `sample` and `log_pdf` blocks inside `_MixtureOfProductDistribution` to apply array masks on continuous variables, isolating `NaN` components from numerical evaluations and preventing calculation warnings.
* **`tests/samplers_tests/tpe_tests/`**: Added a dedicated test suite `test_union_multivariate.py` to confirm convergence on highly nested conditional objective landscapes. Adjusted obsolete fallback warnings in `test_sampler.py`.

### Benchmark Results
Below is the convergence validation evaluated against the hierarchical categorical benchmark script provided in the tracking issue, running 32 repetitions across 150 trials:

conditional_benchmark_results

The benchmark shows that tracking unified conditional parameters using the new `group=False` pipeline prevents early optimization plateaus and establishes standard convergence. Special thanks to @LarsBratholm for proposing the core architecture design and the benchmark script.

<details>
<summary><b>Benchmark script</b></summary>

```python
# Copyright Lars Andersen Bratholm 2024
# Adapted for core TPE Union search space evaluation

import copy
import functools
import warnings
import matplotlib.pyplot as plt
import numpy as np
import optuna
from numpy.typing import NDArray
from optuna.exceptions import ExperimentalWarning
from optuna.samplers import BaseSampler, RandomSampler, TPESampler
from optuna.trial import Trial
from tqdm.contrib.concurrent import process_map


def objective(trial: Trial) -> float:
    x = trial.suggest_categorical("x", [True, False])
    y = trial.suggest_float("y", -1, 1)
    if x is True:
        n = trial.suggest_categorical("n", [True, False])
        if n is True:
            a = trial.suggest_float("a", -1, 1)
            return (a - y) ** 2 + (a + 0.75) ** 2 + 0.025
        b = trial.suggest_float("b", -1, 1)
        return (b - y) ** 2 + (b + 0.25) ** 2 + 0.05
    else:
        m = trial.suggest_categorical("m", [True, False])
        if m is True:
            c = trial.suggest_float("c", -1, 1)
            return (c - y) ** 2 + (c - 0.25) ** 2 + 0.4
        d = trial.suggest_float("d", -1, 1)
        return (d - y) ** 2 + (d - 0.75) ** 2 + 0.01


def run_optimization(n_trials: int, sampler: BaseSampler) -> NDArray[np.float64]:
    sampler_ = copy.deepcopy(sampler)
    study = optuna.create_study(sampler=sampler_, direction="minimize")
    study.enqueue_trial({"x": True, "n": True})
    study.enqueue_trial({"x": True, "n": False})
    study.enqueue_trial({"x": False, "m": True})
    study.enqueue_trial({"x": False, "m": False})

    study.optimize(objective, n_trials=n_trials)
    values = [trial.value for trial in study.trials if trial.value is not None]
    best_values = np.minimum.accumulate(values)
    return best_values


def run_single_benchmark(
    n_trials: int, n_repetitions: int, sampler: BaseSampler, max_workers: int
) -> NDArray[np.float64]:
    results = process_map(
        functools.partial(run_optimization, sampler=sampler),
        (n_trials for _ in range(n_repetitions)),
        max_workers=max_workers,
        desc=sampler.__class__.__name__,
        total=n_repetitions,
        leave=False,
    )
    geometric_mean: NDArray[np.float64] = np.exp(np.sum(np.log(results), axis=0) / n_repetitions)
    return geometric_mean


def run_benchmarks(n_trials: int, n_repetitions: int, max_workers: int) -> None:
    print("Starting benchmarks...")
    r1 = run_single_benchmark(n_trials, n_repetitions, RandomSampler(seed=42), max_workers)
    
    r2 = run_single_benchmark(
        n_trials,
        n_repetitions,
        TPESampler(multivariate=False, n_startup_trials=8, n_ei_candidates=32, seed=42),
        max_workers,
    )
    r3 = run_single_benchmark(
        n_trials,
        n_repetitions,
        TPESampler(multivariate=True, group=True, n_startup_trials=8, n_ei_candidates=256, seed=42),
        max_workers,
    )
    r4 = run_single_benchmark(
        n_trials,
        n_repetitions,
        TPESampler(multivariate=True, group=False, n_startup_trials=8, n_ei_candidates=16, seed=42),
        max_workers,
    )
    
    plt.plot(r1, label="Random Search")
    plt.plot(r2, label="Independent TPE (Core)")
    plt.plot(r3, label="Multivariate TPE (Group=True)")
    plt.plot(r4, label="Multivariate TPE (Group=False)", linewidth=2, linestyle="--")
    
    plt.legend()
    plt.xlabel("Trials")
    plt.ylabel("Loss")
    plt.yscale("log")
    plt.title("Conditional Search Space Convergence Comparison")
    plt.savefig("benchmark.png", dpi=400)
    print("Benchmark complete! Saved results plot to benchmark.png")


if __name__ == "__main__":
    np.random.seed(42)
    optuna.logging.set_verbosity(optuna.logging.ERROR)
    warnings.simplefilter("ignore", category=(UserWarning, ExperimentalWarning))
    run_benchmarks(n_trials=150, n_repetitions=32, max_workers=4)
```

### CODE REVIEW

**Comment by AndreasGerken** [comment]
Thanks a lot for your work on this issue!

**Comment by y0z** [comment]
@not522 @nabenabe0928 Could you review this PR?

**Comment by not522** [comment]
Thank you for working on this. The idea is interesting, but this change may have an impact on Optuna’s core behavior, especially around how conditional parameters are represented and handled in multivariate TPE.

Given that, I think it would be better to first explore this as an OptunaHub package rather than merging it directly into Optuna core. We would be happy to review and discuss it there.

**Comment by Rishabh-git10** [comment]
Thank you for the feedback, @not522. I completely understand the caution regarding modifications to the core TPE codebase.

The main bottleneck with implementing this via `OptunaHub` is that it modifies internal likelihood calculations and array masking inside `ParzenEstimator` and `_MixtureOfProductDistribution`. Moving this to an extension package would require duplicating a significant portion of Optuna's private TPE engine components.

Since the union search space logic is entirely isolated under `multivariate=True` and `group=False`, it carries zero regression risk for default configurations. I am entirely open to either path.

**Comment by not522** [comment]
> The main bottleneck with implementing this via OptunaHub is that it modifies internal likelihood calculations and array masking inside ParzenEstimator and _MixtureOfProductDistribution. Moving this to an extension package would require duplicating a significant portion of Optuna's private TPE engine components.

Duplicate implementations are acceptable in OptunaHub. Copying private implementations is fine.

**Comment by Rishabh-git10** [comment]
Noted. I will port this implementation to the `optunahub-registry`. Closing this PR. Thanks!
